pkgs/development/libraries/webkitgtk/default.nix

```diff
@@ -113,6 +111,7 @@ stdenv.mkDerivation rec {
     at-spi2-core
     enchant2
     libepoxy
+    geoclue2
```

Comment:
You still need to have this conditional

```nix
] ++ lib.optionals (geoclue2 != null) [
  geoclue2
] ++ [
```

since `null` tends to cause issues in inputs.

And at that point, it might still be preferable to use an explicit `enable` argument.

Reply:
I currently arrived at:
```nix
  ++ lib.optional (geoclue2 != null) geoclue2
  ++ lib.optional (libsecret != null) libsecret
```
why can't we just include that as a filter into `stdenv.mkDerivation` (for all the different buildInput arrays)?

Reply:
Adding a filter in `stdenv.mkDerivation` would incur performance cost to everything just for a few instances. Also it might mask passing `null` by accident.

I am not sure what the connection between geolocation and secret service is but would not then a similar equivalence `(geoclue2 != null) ↔(libsecret != null)` exist in the new scheme? I would say that for a feature with multiple dependencies, it is clearer to split out the controlling parameter `assert enableGeolocation -> (geoclue2 != null && libsecret != null)` rather than have `geoclue2` play both roles `assert geoclue2 != null -> libsecret != null`. When properly implemented, it should also be less cumbersome to use – `webkitgtk.override { enableGeolocation = false; }` would disable all geolocation dependencies.

Reply:
> Also it might mask passing `null` by accident.

how does it currently react when passing `null` into `buildInputs`? I was able to "nullify" some attributes of `nixpkgs` in an overlay, and haven't yet hit any assert (besides the one in `webkitgtk`), trace, warning, or such.

Reply:
> ehm, you missed that libsecret and geoclue2 are actually completely independent, the real conflict is between `assert` and `override`/`overrideAttrs` which don't work as expected (e.g. you can pass an updated parameter, but the original parameter at `callPackage` "calltime" after recusion through overlays gets used, so when globally setting `geoclue2=null;` in an overlay, the assert is always hit without any way to work around that)

Oh, the `assert` vs. `override` is a well-known issue, I it just did not occur to me that this is the motivation. It would be useful to mention it in the commit messages. I was also confused by the webkitgtk commit message since it only mentioned geolocation.

> also, I think the option is pretty redundant, especially with the assert in place.

I can see it being redundant when the `geoclue2` is the sole dependency of the feature. I would still prefer not passing it to inputs when it is `null`.

>how does it currently react when passing `null` into `buildInputs`? I was able to "nullify" some attributes of `nixpkgs` in an overlay, and haven't yet hit any assert (besides the one in `webkitgtk`), trace, warning, or such.

It might not be visible when building the package. The issue is that this breaks the expectation that `buildInputs` only contain derivations. Personally, I have seen few breakages in various maintenance and linting tools manipulating packages. And there were also some issues with structured-attrs (though that was with lists within `buildInputs`).

And while filtering would fix the issue for majority of tools (the only exception would be linters that inspect `mkDerivation` arguments like nixpkgs-hammering), allowing `null` would blur the boundary between intentional and accidental. As I understand it, at the moment, even though it is not checked, `null` is not a valid input. If we suddenly allowed that value, we would no longer be able to distinguish between the case when a dependency is removed intentionally and when `null` is passed accidentally.

Allowing to pass `null` to inputs might even be a reasonable thing to do but I would avoid it until there is a consensus that it is a good idea.